specification/dartLangSpec.tex

```diff
@@ -2978,16 +2978,26 @@ \subsubsection{Generative Constructors}
 \end{grammar}
 
 \LMHash{}%
-A \Index{constructor parameter list} is a parenthesized, comma-separated list of formal constructor parameters.
-A \Index{formal constructor parameter} is either a formal parameter (\ref{formalParameters}) or an initializing formal.
-An \Index{initializing formal} has the form \code{\THIS{}.\id}, where \id{} is the name of an instance variable of the immediately enclosing class.
-It is a compile-time error if \id{} is not an instance variable of the immediately enclosing class.
-It is a compile-time error if an initializing formal is used by a function other than a non-redirecting generative constructor.
-
-\LMHash{}%
-If an explicit type is attached to the initializing formal, that is its static type.
-Otherwise, the type of an initializing formal named \id{} is $T_{id}$, where $T_{id}$ is the type of the instance variable named \id{} in the immediately enclosing class.
-It is a compile-time error if the static type of \id{} is not a subtype of $T_{id}$.
+A \Index{constructor parameter list} is
+a parenthesized, comma-separated list of formal constructor parameters.
+A \Index{formal constructor parameter} is a formal parameter
+(\ref{formalParameters}),
+which can be an initializing formal.
+An \Index{initializing formal} contains a term of the form \code{\THIS.\id},
+where \id{} is the name of the initializing formal.
```

Comment:
That sentence sounds descriptive, not prescriptive, but it appears to be defining "initializing formal".

> An \\Index{initializing formal} is a \\syntax{\<fieldFormalParameter\>}. It contains a term of the form \code{\THIS.\id}, where \id{} is the name of the initializing formal.

Reply:
Also: is the grammar for parameters correct. I can't see a production for an initializing function parameter, `(int this.foo(int x))`, but our implementations accept it.

Reply:
(Aha, so I have to reload the page to see comment number 2 and up.)

I was thinking about this sentence, too. We have the grammar rule that introduces initializing formals in a completely different place, `<fieldFormalParameter>` on line 1566.  That rule _does_ allow for the old style function typed parameter (like `int this.foo(int x)`). (It also allows for `var this.foo(int x)` that I've recommended that we turn into an error ;-).

But the point is that (presumably) the text around line 2984 used to refer to a grammar rule that actually had an alternative for "normal" parameters and another alternative for initializing formal parameters. And then the grammar was updated, some time around 1812 (OK, 2012 maybe), but the text right here wasn't updated.

This means that it's now a non-syntax error to have an initializing formal in any other function than a generative non-redirecting constructor, and that error has been added as it should.

But it's confusing that we're not describing the properties of initializing formals at the same location where the grammar is specified.

Reply:
Ah, yes:
> ```
>  fieldFormalParameter> ::= \gnewline{}
>       <finalConstVarOrType>? \THIS{} `.' <identifier> <formalParameterPart>?
>```
which allows parameters afterwards. It's missing the `?`? to allow for a nullable function type, but otherwise it's fine.
(So, it should be:

> ```
> <fieldFormalParameter> ::= \gnewline{}
>       <finalConstVarOrType>? \THIS{} `.' <identifier> (<formalParameterPart> `?'?)?
>```

Reply:
Uploaded a commit which is considerably clearer on these things. Also, it moves the error for having an init. formal to the section about parameters, with commentary referring to it in the constructor section. After all, this is a rule that everybody _else_ than those constructors need to follow.

Reply:
We got rid of `var` a while ago via a separate compile-time error, in the text below. In order to rock the boat a bit less, I've proposed a new grammar rule that keeps the `<finalConstVarOrType>` and just adds the optional question mark at the end of a parameter list.

Reply:
@lrhn, PTAL.